.github/scripts/checkstyle.py

```diff
@@ -0,0 +1,98 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from glob import glob
+import logging
+import os
+import os.path
+import sys
+from typing import Tuple, Optional
+import xml.etree.ElementTree
+
+
+logger = logging.getLogger()
+logger.setLevel(logging.DEBUG)
+handler = logging.StreamHandler(sys.stderr)
+handler.setLevel(logging.DEBUG)
+logger.addHandler(handler)
+
+
+def get_env(key: str) -> str:
+    value = os.getenv(key)
+    logger.debug(f"Read env {key}: {value}")
+    return value
+
+
+def parse_report(workspace_path, fp) -> Tuple[int, int]:
+    stack = []
+    errors = []
+    file_count = 0
+    error_count = 0
+    for (event, elem) in xml.etree.ElementTree.iterparse(fp, events=["start", "end"]):
+        if event == "start":
+            stack.append(elem)   
+            if elem.tag == "file":
+                file_count += 1
+                errors.clear()
+            if elem.tag == "error":
+                logger.debug(f"Found checkstyle error: {elem.attrib}")
+                errors.append(elem)
+                error_count += 1
+        elif event == "end":
+            if elem.tag == "file" and len(errors) > 0:
+                filename = elem.get("name")
+                rel_path = os.path.relpath(filename, workspace_path)
+                logger.debug(f"Outputting errors for file: {elem.attrib}")
+                for error in errors:
+                    line = error.get("line")
+                    col = error.get("column")
+                    severity = error.get("severity")
+                    message = error.get('message')
+                    title = f"Checkstyle {severity}"
+                    print(f"::notice file={rel_path},line={line},col={col},title={title}::{message}")
+            stack.pop()
+        else:
+            logger.error(f"Unhandled xml event {event}: {elem}")
+    return file_count, error_count
+
+
+if __name__ == "__main__":
+    """
+    Parse checkstyle XML reports and generate GitHub annotations.
```

Comment:
I do love this script, but have you considered using existent Github action? like https://github.com/jwgmeligmeyling/checkstyle-github-action

Reply:
Yes, I did look through the existing marketplace actions before writing this script. We must be cautious when bringing in 3rd party actions (actions from non-verified authors). ASF Infra team has to manually audit each 3rd party action that gets included (similar to the review process for a Jenkins plugin). There was one verified action https://github.com/marketplace/check-run-reporter which seems to cost money. 

Going the custom route is also based on my experience when trying to process JUnit tests with a marketplace action -- it wasn't able to handle the number of tests we have (kept running out of memory).

I figured this was a simple enough task that we could just roll our own script 😄

Reply:
@mumrah thanks for sharing. I guess we need to have custom parser for spotbug report too, right?

Reply:
Yes, and JUnit eventually. I'll take a look at those next. I suspect it will be quite similar to the checkstyle parser (glob the right path, streaming xml, etc).